### PR TITLE
Descheduler profiles

### PR DESCRIPTION
**Pull request by damemi**

### CODE REVIEW

File: enhancements/scheduling/descheduler-profiles.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  - "@soltysh"
+  - "@ingvagabund"
+approvers:
+  - "@soltysh"
+  - "@ingvagabund"
+creation-date: 2020-11-23
+last-updated: 2020-11-23
+status: provisional
+see-also:
+  - "https://github.com/openshift/enhancements/pull/541/"
```
Comment (soltysh) [inline enhancements/scheduling/descheduler-profiles.md:15]: I don't expect the other PR not merging, so I'd just link to the file like below.

File: enhancements/scheduling/descheduler-profiles.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  - "/enhancements/kube-apiserver/audit-policy.md"
+replaces:
+superseded-by:
+---
+
+# Descheduler Profiles
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
```
Comment (soltysh) [inline enhancements/scheduling/descheduler-profiles.md:25]: Check this, we are implementing this right away.

File: enhancements/scheduling/descheduler-profiles.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+This approach will clearly present users with all of their options for configuration
+while eliminating the chance of typos and the need for validation against input.
+
+An alternative would be to implement this strictly as a list:
+```go
+type KubeDeschedulerSpec struct {
+  Profiles []string `json:"profiles"`
+  ...
+}
```
Comment (soltysh) [inline enhancements/scheduling/descheduler-profiles.md:138]: Another alternative is a list of enums.

File: enhancements/scheduling/descheduler-profiles.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+grouping some strategies into their own profiles.
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+Ideally, this field will immediately replace the existing `strategies` config as the initial GA option for
+configuring the descheduler.
+
+However, since the current API is version `v1beta1` we may be required to continue supporting the
```
Comment (soltysh) [inline enhancements/scheduling/descheduler-profiles.md:229]: I'd clearly express the deprecation timeline, which is Beta: 9 months or 3 releases (whichever is longer), for beta-level APIs. For conversions you'll need to have both (`strategies` and `profile`) added to both APIs, but v1beta1 version should be deprecated right away, same with `strategies` field in v1, and point to using the new `profiles`.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: enhancements/scheduling/descheduler-profiles.md (hunk 3 above)
Comment (damemi) [reply enhancements/scheduling/descheduler-profiles.md:138]: are there any opinions on preference for either one? I'm not attached either way but curious if anyone else had any insight

File: enhancements/scheduling/descheduler-profiles.md (hunk 4 above)
Comment (damemi) [reply enhancements/scheduling/descheduler-profiles.md:229]: ok, so for the v1 API we should still provide strategies with a note that it's deprecated and will be removed? Or just provide the v1 API with only `profiles` and still support the v1beta1 api with `strategies` until that API is removed according to that timeline?

**Review by damemi — COMMENTED** [review]
updated with feedback and a couple more questions

File: enhancements/scheduling/descheduler-profiles.md (hunk 4 above)
Comment (soltysh) [reply enhancements/scheduling/descheduler-profiles.md:229]: @damemi yeah, for v1 provide deprecated strategies. We should also announce deprecation of v1beta1 too.

File: enhancements/scheduling/descheduler-profiles.md (hunk 3 above)
Comment (soltysh) [reply enhancements/scheduling/descheduler-profiles.md:138]: let's go with a list of enums, we can then re-use the kubebuilder validation for enums for that type. The only thing we'll have to ensure is to de-duplication or call out what happens and validation if contradicting profiles are applied.

**Comment by damemi** [comment]
To clarify on what we decided for this, we will add both fields to v1beta1 and remove functionality for the old field (only providing it for transition support). When v1beta1 is removed in favor of v1, those fields will be removed and not present in v1. I have updated the enhancement to include this.

**Review by soltysh — APPROVED** [review]
This was discussed earlier today during our call and this was agreed between @damemi, @ingvagabund, @deads2k and myself

/lgtm
/approve

**Pull request merged**